cmd/sanssh/client/completion.go

```diff
@@ -0,0 +1,227 @@
+}
+
+type subCompleter struct {
+	command subcommands.Command
+}
+
+func (c *subCompleter) ListCommandsAndEssentialFlags() []string {
+	s, ok := c.command.(client.HasSubpackage)
+	if !ok {
+		return nil
+	}
+	return fromSubpackage(s).ListCommandsAndEssentialFlags()
+}
+func (c *subCompleter) GetCommandCompleter(cmd string) completer {
+	s, ok := c.command.(client.HasSubpackage)
+	if !ok {
+		return nil
+	}
+	return fromSubpackage(s).GetCommandCompleter(cmd)
+}
+func (c *subCompleter) ListFlags() []string {
+	s := flag.NewFlagSet(c.command.Name(), flag.ContinueOnError)
+	c.command.SetFlags(s)
+	var flags []string
+	s.VisitAll(func(f *flag.Flag) {
+		flags = append(flags, "--"+f.Name)
+	})
+	return flags
+}
+
+func (c *subCompleter) GetFlag(flag, prefix string) []string {
+	return []string{prefix}
+}
+
+func (c *subCompleter) IsBoolFlag(fl string) bool {
+	s := flag.NewFlagSet(c.command.Name(), flag.ContinueOnError)
+	c.command.SetFlags(s)
+	return flagIsBool(fl, s.VisitAll)
+}
+
+func (c *subCompleter) GetArgs(prefix string) []string {
+	if args, ok := c.command.(client.PredictArgs); ok {
+		return args.PredictArgs(prefix)
+	}
+	return nil
+}
+
+func filterToPrefix(args []string, prefix string) []string {
+	var s []string
+	for _, a := range args {
+		if strings.HasPrefix(a, prefix) {
+			s = append(s, a)
+		}
+	}
+	return s
+}
+
+var validBools = map[string]bool{
+	"1": true, "0": true, "t": true, "f": true, "T": true, "F": true, "true": true, "false": true, "TRUE": true, "FALSE": true, "True": true, "False": true,
+}
+
+func predict(c completer, args []string) []string {
+	// Flags require special handling
+	if strings.HasPrefix(args[0], "-") {
+		flagName := strings.TrimLeft(args[0], "-")
+		switch len(args) {
+		case 1:
+			return filterToPrefix(c.ListFlags(), args[0])
+		case 2:
+			if c.IsBoolFlag(flagName) {
+				return predict(c, args[1:])
+			}
+			return c.GetFlag(flagName, args[1])
+		default:
+			if strings.Contains(flagName, "=") || (c.IsBoolFlag(flagName) && !validBools[args[1]]) {
+				return predict(c, args[1:])
+			}
+			return predict(c, args[2:])
+		}
+	}
+	if len(args) > 1 {
+		next := c.GetCommandCompleter(args[0])
+		if next == nil {
+			prefix := strings.Join(args, " ")
+			return filterToPrefix(c.GetArgs(prefix), prefix)
+		}
+		return predict(next, args[1:])
+	}
+	return filterToPrefix(c.ListCommandsAndEssentialFlags(), args[0])
+}
+
+// Predictor can provide suggested values based on a provided prefix.
+type Predictor func(prefix string) []string
+
+// AddCommandLineCompletion adds command line completion for shells. It should be called after
+// registering all subcommands and before calling flag.Parse().
+//
+// topLevelFlagPredictions gives optional finer-grained control over predictions on the top-level
+// flags.
+func AddCommandLineCompletion(topLevelFlagPredictions map[string]Predictor) {
+	// If COMP_LINE is present, we're running as command completion.
+	compLine, ok := os.LookupEnv("COMP_LINE")
```

Comment:
out of curiosity, would it be possible to implement this in the same basic way that 'help' and 'commands' work in 'subcommands' (i.e. as just another subcommand)?

that is, instead of `COMP_LINE="--" ./sanssh file foo`, could you implement it as `./sanssh complete <args>`

Reply:
It'd require writing a bit of bash that I'd like to avoid. COMP_LINE is one of the defined variables for the `complete` command, so avoiding the environmental variable here would likely mean creating a bash function that invokes `./sanssh complete $COMP_LINE`.

I've been reading through "Programmable Completion" in https://linux.die.net/man/1/bash to figure out what's possible.

Reply:
Resolving because I like the environmental variable more than a bash wrapper.